pkg/apis/core/validation/validation.go

```diff
@@ -5137,6 +5137,46 @@ func ValidateContainerStateTransition(newStatuses, oldStatuses []core.ContainerS
 	return allErrs
 }
 
+// ValidateInitContainerStateTransition test to if any illegal init container state transitions are being attempted
+func ValidateInitContainerStateTransition(newStatuses, oldStatuses []core.ContainerStatus, fldpath *field.Path, podSpec *core.PodSpec) field.ErrorList {
+	allErrs := field.ErrorList{}
+	// If we should always restart, containers are allowed to leave the terminated state
+	if podSpec.RestartPolicy == core.RestartPolicyAlways {
+		return allErrs
+	}
+	for i, oldStatus := range oldStatuses {
+		// Skip any container that is not terminated
+		if oldStatus.State.Terminated == nil {
+			continue
+		}
+		// Skip any container that failed but is allowed to restart
+		if oldStatus.State.Terminated.ExitCode != 0 && podSpec.RestartPolicy == core.RestartPolicyOnFailure {
+			continue
+		}
+
+		// Skip any restartable init container that is allowed to restart
+		isRestartableInitContainer := false
+		for _, c := range podSpec.InitContainers {
+			if oldStatus.Name == c.Name {
+				if c.RestartPolicy != nil && *c.RestartPolicy == core.ContainerRestartPolicyAlways {
```

Comment:
should we also check the feature gate here? what happens if the feature is disabled after the pod was created?

Reply:
I think we shouldn't.

> what happens if the feature is disabled after the pod was created?

It ends up breaking the existing pods having sidecar containers if we check the feature gate here.

Reply:
well, we do check the feature gate in other places?

Reply:
https://github.com/kubernetes/kubernetes/blob/109491f3cbaefcaf95e9e94ea0c216bfc38d6570/pkg/api/pod/util.go#L589-L595

Yes, the apiserver drops container.restartPolicy field if the FG is disabled AND the existing pod has no sidecar.

Reply:
apiserver should allows the update requests for the existing sidecar container.

Reply:
To be clear, if the container.restartPolicy is not nil (not dropped) here, it means that the feature gate is enabled or this request is updating the existing sidecar container which should be allowed.

Reply:
We don't usually want to check gates except at "admission", or else we allow data in and then fail updates.  That said, I would drop a comment here with the gate name so that when we eventually remove the gate, we can grep for it and find this and remove the nil check.